src/Microsoft.TestPlatform.AdapterUtilities/ManagedNameUtilities/ManagedNameHelper.Reflection.cs

```diff
@@ -266,14 +268,169 @@ private static void AppendTypeString(StringBuilder b, Type type, bool closedType
             }
         }
 
-        private static void AppendNestedTypeName(StringBuilder b, Type type)
+        private static void AppendNamespace(StringBuilder b, string ns)
+        {
+            int start = 0;
+            bool shouldEscape = false;
+
+            for (int i = 0; i <= ns.Length; i++)
+            {
+                if (i == ns.Length || ns[i] == '.')
+                {
+                    if (start != 0)
+                    {
+                        b.Append('.');
+                    }
+
+                    var part = ns.Substring(start, i - start);
+                    if (shouldEscape)
+                    {
+                        NormalizeString(b, part);
+                        shouldEscape = false;
+                    }
+                    else
+                    {
+                        b.Append(part);
+                    }
+
+                    start = i + 1;
+                    continue;
+                }
+
+                shouldEscape = shouldEscape || NeedsEscaping(ns[i], i - start);
+            }
+        }
+
+        private static void AppendMethodString(StringBuilder methodBuilder, string name, int methodArity)
```

Comment:
Some documentation explaining the business logic of this method would be nice since it doesn't do trivial work.

Unit tests would be nice too.